File: architecture-skills/design-patterns/examples/factory-method/factory-method.swift
// Factory Method pattern in Swift.
//
// The creator is a protocol whose extension holds the shared workflow;
// conforming types supply only the factory method. This is the idiomatic
// Swift replacement for an abstract base class.

import Foundation

protocol Transport {
    func deliver(_ cargo: String) -> String
}

struct Truck: Transport {
    func deliver(_ cargo: String) -> String {
        "Truck delivers \(cargo) by road in a box"
    }
}

struct Ship: Transport {
    func deliver(_ cargo: String) -> String {
        "Ship delivers \(cargo) by sea in a container"
    }
}

/// Creator: conforming types supply only the factory method.
protocol Logistics {
    var name: String { get }
    func createTransport() -> Transport
}

extension Logistics {
    /// Shared workflow, written once for every creator.
    func planDelivery(of cargo: String) -> String {
        let transport = createTransport()
        return "[\(name)] \(transport.deliver(cargo))"
    }
}

struct RoadLogistics: Logistics {
    let name = "RoadLogistics"
    func createTransport() -> Transport { Truck() }
}

struct SeaLogistics: Logistics {
    let name = "SeaLogistics"
    func createTransport() -> Transport { Ship() }
}

let all: [Logistics] = [RoadLogistics(), SeaLogistics()]
for logistics in all {
    print(logistics.planDelivery(of: "machine parts"))
}

File: architecture-skills/design-patterns/examples/factory-method/factory-method.go
// Factory Method pattern in Go.
//
// Go has no inheritance, so the shared workflow is a plain function that
// accepts a creator interface (composition over subclassing). Concrete
// creators are small structs that supply only the factory method.

package main

import "fmt"

type Transport interface {
	Deliver(cargo string) string
}

type truck struct{}

func (truck) Deliver(cargo string) string {
	return "Truck delivers " + cargo + " by road in a box"
}

type ship struct{}

func (ship) Deliver(cargo string) string {
	return "Ship delivers " + cargo + " by sea in a container"
}

// Logistics is the creator: it supplies only the factory method.
type Logistics interface {
	Name() string
	CreateTransport() Transport
}

type RoadLogistics struct{}

func (RoadLogistics) Name() string               { return "RoadLogistics" }
func (RoadLogistics) CreateTransport() Transport { return truck{} }

type SeaLogistics struct{}

func (SeaLogistics) Name() string               { return "SeaLogistics" }
func (SeaLogistics) CreateTransport() Transport { return ship{} }

// PlanDelivery is the shared workflow, written once against the interface.
func PlanDelivery(logistics Logistics, cargo string) string {
	transport := logistics.CreateTransport()
	return "[" + logistics.Name() + "] " + transport.Deliver(cargo)
}

func main() {
	for _, logistics := range []Logistics{RoadLogistics{}, SeaLogistics{}} {
		fmt.Println(PlanDelivery(logistics, "machine parts"))
	}
}
